### PR TITLE
[profiler][refactor] Avoid using legacy event in profiler

### PR DESCRIPTION
**Pull request by ilia-cher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61931
* __->__ #61721

Summary:
Remove dependency on LegacyEvent from the profiler

Test Plan:
python test/test_profiler.py -v

Reviewers: gdankel

Differential Revision: [D29716769](https://our.internmc.facebook.com/intern/diff/D29716769)

### CODE REVIEW

**Comment by ilia-cher** [comment]
@ilia-cher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29716769).

**Comment by ilia-cher** [comment]
@ilia-cher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29716769).

**Comment by ilia-cher** [comment]
@ilia-cher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29716769).

**Comment by ilia-cher** [comment]
@ilia-cher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29716769).

File: torch/autograd/profiler.py [hunk 1]
```diff
@@ -1148,9 +1149,9 @@ def parse_kineto_results(result):
             for mem_record in mem_records:
                 if (mem_record[0].start_us() >= kineto_event.start_us() and
                         mem_record[0].start_us() <= abs_end_us):
```
Comment (guotuofeng) [inline torch/autograd/profiler.py:1151]: this piece of code has performance penalty as we tested. 

https://github.com/pytorch/pytorch/issues/58983 
https://github.com/pytorch/pytorch/issues/61041
https://github.com/pytorch/kineto/issues/308 are complaining this performance issue when handling memory events.

https://github.com/pytorch/pytorch/pull/60432 improve the performance of memory paring

Comment (ilia-cher) [reply torch/autograd/profiler.py:1151]: I'll rebase after your fix lands

**Comment by ilia-cher** [comment]
@ilia-cher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29716769).

**Comment by ilia-cher** [comment]
@ilia-cher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29716769).

**Comment by ilia-cher** [comment]
@ilia-cher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29716769).

File: torch/csrc/autograd/profiler_kineto.cpp [hunk 2]
```diff
@@ -529,6 +518,10 @@ int64_t KinetoEvent::cudaElapsedUs() const {
 }
 
 c10::DeviceType KinetoEvent::deviceType() const {
+  if (device_type_ >= 0) {
```
Comment (kimishpatel) [inline torch/csrc/autograd/profiler_kineto.cpp:521]: Why make this `int8_t` and defaulted `-1`? Can you add invalid device type? Or this is tied, 1:1, to device types on aten side so we have to follow that?

File: torch/csrc/autograd/profiler_kineto.cpp [hunk 3]
```diff
@@ -529,6 +518,10 @@ int64_t KinetoEvent::cudaElapsedUs() const {
...
+    return (c10::DeviceType)device_type_;
```
Comment (kimishpatel) [inline torch/csrc/autograd/profiler_kineto.cpp:522]: Ok ignore my earlier comment. I see you cast to `c10::DeviceType`

File: torch/csrc/autograd/profiler_kineto.cpp [hunk 4]
```diff
@@ -182,16 +180,11 @@ struct KinetoThreadLocalState : public ProfilerThreadLocalState {
...
-    cpu_trace->activities.insert(
-      cpu_trace->activities.end(),
-      memory_events_.begin(),
-      memory_events_.end());
   }
 
+  uint64_t start_time_;
   std::vector<KinetoEvent> kineto_events_;
   std::unique_ptr<libkineto::CpuTraceBuffer> cpu_trace;
-  std::vector<libkineto::GenericTraceActivity> memory_events_;
```
Comment (kimishpatel) [inline torch/csrc/autograd/profiler_kineto.cpp:194]: Does this mean that support for memory profiling is removed?

Comment (ilia-cher) [reply torch/csrc/autograd/profiler_kineto.cpp:194]: no it doesn't, we pass the memory events in kineto_events_ (as we should've)

File: torch/autograd/profiler.py [hunk 5]
```diff
@@ -1226,20 +1229,21 @@ def parse_kineto_results(result):
...
                 start_us=rel_start_us,
                 end_us=rel_start_us,  # no duration
-                fwd_thread=mem_record[0].fwd_thread_id(),
+                fwd_thread=mem_record[0].start_thread_id(),
                 input_shapes=[],
                 stack=[],
-                scope=mem_record[0].scope(),
-                cpu_memory_usage=mem_record[0].cpu_memory_usage(),
-                cuda_memory_usage=mem_record[0].cuda_memory_usage(),
+                scope=0,  # function
```
Comment (kimishpatel) [inline torch/autograd/profiler.py:1244]: Can you add a comment to point to `aten/src/ATen/record_function.h` where RecordScope is and 0 = Function?

File: torch/autograd/profiler.py [hunk 6]
```diff
@@ -1134,29 +1135,29 @@ def rewrite_name(name, with_wildcard=False):
...
-    mem_records = []
-    for record in itertools.chain(*result.legacy_events()):
-        if record.kind() == 'mark' and record.name() == '__start_profile':
-            assert start_record is None
-            start_record = record
-        if record.kind() == 'memory_alloc':
-            mem_records.append([record, False])
-    assert start_record is not None, "Invalid profiler output, __start_profile is missing"
+    trace_start_us = result.trace_start_us()
+    mem_records = [[evt, False] for evt in result.events() if evt.name() == "[memory]"]
```
Comment (kimishpatel) [inline torch/autograd/profiler.py:1140]: What does 'False' mean in `[evt, False]`?

**Review by kimishpatel — APPROVED** [review]
Looks good to me. Left some minor comments.

File: torch/autograd/profiler.py (hunk 6 above)
Comment (ilia-cher) [reply torch/autograd/profiler.py:1140]: it's a flag that is used later in the code to filter out "top-level" memory events (events that did not happen in the context of another op), that are outputted separately

**Comment by ilia-cher** [comment]
@ilia-cher has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29716769).

**Pull request closed without merging**